File: python/othello/board.py
"""
Class Board
Defines the game board
Akseli Lukkarila
2019-2026
"""

try:
    from othello.colorprint import Color, get_color, print_yellow
    from othello.models import Direction, Disk, Move, Square, Step
except ModuleNotFoundError:
    from colorprint import Color, get_color, print_yellow
    from models import Direction, Disk, Move, Square, Step

UP = 1
DOWN = -1
LEFT = -1
RIGHT = 1
STILL = 0

# All possible step directions for a square on the board.
STEP_DIRECTIONS = (
    Step(DOWN, LEFT),
    Step(DOWN, RIGHT),
    Step(DOWN, STILL),
    Step(STILL, LEFT),
    Step(STILL, RIGHT),
    Step(UP, LEFT),
    Step(UP, RIGHT),
    Step(UP, STILL),
)


class Board:
    """Handles game board state and logic."""

    def __init__(self, size=8):
        """Initialize a new board for the given board size."""
        board = self._init_board(size)
        # Index list (0...size) to avoid repeating same range in loops.
        # Not really needed in Python but kept here to more closely match other implementations...
        indices = tuple(range(size))

        # Keep track of empty squares on board to avoid checking already filled positions.
        empty_squares = self._init_empty_squares(size, board)

        self._board: list[Disk] = board
        self._empty_squares: set[Square] = empty_squares
        self._indices: tuple[int, ...] = indices
        self._size: int = size

    def can_play(self) -> bool:
        """Return true if board contains empty squares."""
        return bool(self._empty_squares)

    def place_disk(self, chosen_move: Move) -> None:
        """
        Update board for given disk placement.
        Throws AssertionError if the move is not a valid move.
        """
        start = chosen_move.square
        assert self._check_square(start), f"Invalid coordinates: {start}"
        assert self._get_square(start) == Disk.EMPTY, (
            f"Trying to place disk to an occupied square: {start}!"
        )
        self._set_square(start, chosen_move.disk)
        self._empty_squares.remove(start)
        for square in chosen_move.affected_squares():
            self._set_square(square, chosen_move.disk)

    def possible_moves(self, disk: Disk) -> list[Move]:
        """Returns a list of possible moves for the given player."""
        moves = []
        opposing_disk = disk.opponent()
        for square in self._empty_squares:
            value = 0
            directions = []
            for step in STEP_DIRECTIONS:
                pos = square + step
                # Next square in this direction needs to be the opposing disk
                if self._get_square(pos) != opposing_disk:
                    continue

                num_steps = 0
                # Keep stepping over opponents disks
                while self._get_square(pos) == opposing_disk:
                    num_steps += 1
                    pos += step
                # Valid move only if a line of opposing disks ends with own disk
                if self._get_square(pos) == disk:
                    directions.append(Direction(step, num_steps))
                    value += num_steps

            if value:
                moves.append(Move(square, disk, value, directions))

        return sorted(moves)

    def print_possible_moves(self, moves: list[Move]) -> None:
        """Print board with available move coordinates and the resulting points gained."""
        print_yellow(f"  Possible moves ({len(moves)}):")
        # Convert board from Disk enums to strings
        formatted_board = [disk.board_char_with_color() for disk in self._board]
        # Add possible moves to board
        for possible_move in moves:
            index = self._square_index(possible_move.square)
            formatted_board[index] = get_color(str(possible_move.value), Color.yellow)
            print(f"  {possible_move}")

        # Print board with move positions
        header = " ".join(["    "] + [get_color(str(i), bold=True) for i in self._indices])
        board = "\n".join(
            f"  {get_color(str(y), bold=True)} "
            + " ".join(formatted_board[y * self._size + x] for x in self._indices)
            for y in self._indices
        )
        print(f"{header}\n{board}")

    def print_score(self) -> None:
        """Print current score for both players."""
        black, white = self._player_scores()
        print(f"\n{self}")
        print(
            f"Score: {get_color(str(black), Color.magenta)} | {get_color(str(white), Color.cyan)}"
        )

    def result(self) -> Disk:
        """Returns the winning disk colour. Empty indicates a draw."""
        match self._score():
            case 0:
                return Disk.EMPTY
            case score if score > 0:
                return Disk.WHITE
            case _:
                return Disk.BLACK

    def log_entry(self) -> str:
        """Get board status string for game log."""
        return "".join(disk.board_char() for disk in self._board)

    def _check_coordinates(self, x: int, y: int) -> bool:
        """Check that the given coordinates are valid (inside the board)."""
        return 0 <= x < self._size and 0 <= y < self._size

    def _check_square(self, square: Square) -> bool:
        """Check that the given square is valid (inside the board)."""
        return self._check_coordinates(square.x, square.y)

    def _get_square(self, square: Square) -> Disk | None:
        """Returns the state of the board (empty, white, black) at the given square."""
        if self._check_square(square):
            index = self._square_index(square)
            return self._board[index]

        return None

    def _square_index(self, square: Square) -> int:
        """Map square to board index."""
        return square.y * self._size + square.x

    def _player_scores(self) -> tuple[int, int]:
        """Count and return the number of black and white disks."""
        # Alternative that is more readable but iterates the board twice:
        # black = sum(disk == Disk.BLACK for disk in self._board)
        # white = sum(disk == Disk.WHITE for disk in self._board)
        black = 0
        white = 0
        for disk in self._board:
            if disk == Disk.BLACK:
                black += 1
            elif disk == Disk.WHITE:
                white += 1

        return black, white

    def _score(self) -> int:
        """
        Returns the total score.
        Positive value means more white disks and negative means more black disks.
        """
        return sum(disk.value for disk in self._board)

    def _set_square(self, square: Square, disk: Disk) -> None:
        """Sets the given square to the given value."""
        assert self._check_square(square), f"Invalid coordinates: {square}"
        index = self._square_index(square)
        self._board[index] = disk

    @staticmethod
    def _init_board(size: int) -> list[Disk]:
        """Initialize game board with starting disk positions."""
        # Initialize game board with empty disks
        board = [Disk.EMPTY for _ in range(size * size)]
        # Set starting positions
        row = (size - 1) // 2 if size % 2 == 0 else (size - 1) // 2 - 1
        col = size // 2
        board[row * size + row] = Disk.WHITE
        board[row * size + col] = Disk.BLACK
        board[col * size + row] = Disk.BLACK
        board[col * size + col] = Disk.WHITE
        return board

    @staticmethod
    def _init_empty_squares(size: int, board: list[Disk]) -> set[Square]:
        """Initialize empty squares for the board."""
        return {
            Square(x, y)
            for x in range(size)
            for y in range(size)
            if board[y * size + x] == Disk.EMPTY
        }

    def __str__(self) -> str:
        """Format game board to string."""
        # Horizontal indices
        board = f"  {get_color(' '.join(str(i) for i in self._indices), bold=True)}"
        for y in self._indices:
            # Vertical index
            board += f"\n{get_color(str(y), bold=True)}"
            # Row values
            for x in self._indices:
                board += f" {self._board[y * self._size + x].board_char_with_color()}"

        return board

File: swift/Sources/OthelloLib/board.swift
//==========================================================
// Board
// Defines the game board
// Akseli Lukkarila
// 2019-2026
//==========================================================

import ColorizeSwift

private let UP = 1
private let DOWN = -1
private let LEFT = -1
private let RIGHT = 1
private let STILL = 0

/// All possible step directions for a square on the board.
private let STEP_DIRECTIONS: [Step] = [
    Step(x: DOWN, y: LEFT),
    Step(x: DOWN, y: RIGHT),
    Step(x: DOWN, y: STILL),
    Step(x: STILL, y: LEFT),
    Step(x: STILL, y: RIGHT),
    Step(x: UP, y: LEFT),
    Step(x: UP, y: RIGHT),
    Step(x: UP, y: STILL),
]

/// Handles game board state and logic.
class Board {
    private var board: [Disk]
    private var emptySquares: Set<Square>
    private let indices: [Int]
    let size: Int

    /// Initialize a new board for the given board size.
    init(size: Int) {
        let board = Self.initBoard(size)
        // Index list (0...size) to avoid repeating same range in loops.
        let indices = Array(0 ..< size)

        // Keep track of empty squares on board to avoid checking already filled positions.
        let emptySquares = Self.initEmptySquares(size, board)

        self.board = board
        self.emptySquares = emptySquares
        self.indices = indices
        self.size = size
    }

    /// Return true if board contains empty squares.
    func canPlay() -> Bool {
        !self.emptySquares.isEmpty
    }

    /// Update board for given disk placement.
    func placeDisk(_ chosenMove: Move) {
        let start = chosenMove.square
        guard let disk = self.getSquare(start) else {
            preconditionFailure("Invalid coordinates: \(start)")
        }
        precondition(disk == .empty, "Trying to place disk to an occupied square: \(start)!")
        self.setSquare(start, chosenMove.disk)
        self.emptySquares.remove(start)
        for square in chosenMove.affectedSquares() {
            self.setSquare(square, chosenMove.disk)
        }
    }

    /// Returns a list of possible moves for the given player.
    func possibleMoves(disk: Disk) -> [Move] {
        var moves: [Move] = []
        let opposingDisk = disk.opponent()
        for square in self.emptySquares {
            var value = 0
            var directions: [Direction] = []
            for step in STEP_DIRECTIONS {
                var pos = square + step
                // Next square in this direction needs to be the opposing disk
                if self.getSquare(pos) != opposingDisk {
                    continue
                }
                var numSteps = 0
                // Keep stepping over opponents disks
                while self.getSquare(pos) == opposingDisk {
                    numSteps += 1
                    pos += step
                }
                // Valid move only if a line of opposing disks ends with own disk
                if self.getSquare(pos) == disk {
                    directions.append(Direction(step: step, count: numSteps))
                    value += numSteps
                }
            }
            if value > 0 {
                moves.append(Move(square: square, disk: disk, value: value, directions: directions))
            }
        }
        moves.sort()
        return moves
    }

    /// Print board with available move coordinates and the resulting points gained.
    func printPossibleMoves(_ moves: [Move]) {
        printYellow("  Possible moves (\(moves.count)):")
        // Convert board from Disk enums to strings
        var formattedBoard = self.board.map { $0.boardCharWithColor() }

        // Add possible moves to board
        for possibleMove in moves {
            let index = self.squareIndex(possibleMove.square)
            formattedBoard[index] = "\(possibleMove.value)".yellow()
            print("  \(possibleMove)")
        }
        // Print board with move positions
        let header = (["    "] + self.indices.map { "\($0)".bold() })
            .joined(separator: " ")

        let board = self.indices
            .map { y in
                let row = self.indices
                    .map { x in formattedBoard[y * self.size + x] }
                    .joined(separator: " ")
                return "  \("\(y)".bold()) \(row)"
            }
            .joined(separator: "\n")

        print("\(header)\n\(board)")
    }

    /// Print current score for both players.
    func printScore() {
        let (black, white) = self.playerScores()
        print("\n\(self)")
        print(
            "Score: \(getColor(black, TerminalStyle.magenta)) | \(getColor(white, TerminalStyle.cyan))"
        )
    }

    /// Returns the winning disk colour. Empty indicates a draw.
    func result() -> Disk {
        let totalScore = self.score()
        if totalScore > 0 {
            return .white
        }
        if totalScore < 0 {
            return .black
        }
        return .empty
    }

    /// Get board status string for game log.
    func logEntry() -> String {
        self.board.map { $0.boardChar() }.joined()
    }

    /// Check that the given coordinates are valid (inside the board).
    private func checkCoordinates(_ x: Int, _ y: Int) -> Bool {
        x >= 0 && x < self.size && y >= 0 && y < self.size
    }

    /// Check that the given square is valid (inside the board).
    private func checkSquare(_ square: Square) -> Bool {
        self.checkCoordinates(square.x, square.y)
    }

    /// Returns the state of the board (empty, white, black) at the given square.
    func getSquare(_ square: Square) -> Disk? {
        if self.checkSquare(square) {
            let index = self.squareIndex(square)
            return self.board[index]
        }
        return nil
    }

    /// Map square to board index.
    private func squareIndex(_ square: Square) -> Int {
        square.y * self.size + square.x
    }

    /// Count and return the number of black and white disks.
    func playerScores() -> (black: Int, white: Int) {
        var black = 0
        var white = 0
        for disk in self.board {
            switch disk {
                case .black:
                    black += 1
                case .white:
                    white += 1
                case .empty:
                    break
            }
        }
        return (black, white)
    }

    /// Returns the total score.
    /// Positive value means more white disks and negative means more black disks.
    func score() -> Int {
        self.board.reduce(0) { $0 + $1.rawValue }
    }

    /// Sets the given square to the given value.
    private func setSquare(_ square: Square, _ disk: Disk) {
        precondition(self.checkSquare(square), "Invalid coordinates: \(square)")
        let index = self.squareIndex(square)
        self.board[index] = disk
    }

    /// Initialize game board with starting disk positions.
    private static func initBoard(_ size: Int) -> [Disk] {
        // Initialize game board with empty disks
        var board = [Disk](repeating: .empty, count: size * size)
        // Set starting positions
        let row = size % 2 == 0 ? (size - 1) / 2 : (size - 1) / 2 - 1
        let col = size / 2
        board[row * size + row] = .white
        board[row * size + col] = .black
        board[col * size + row] = .black
        board[col * size + col] = .white
        return board
    }

    /// Initialize empty squares for the board.
    private static func initEmptySquares(_ size: Int, _ board: [Disk]) -> Set<Square> {
        var emptySquares = Set<Square>()
        for index in 0 ..< board.count where board[index] == .empty {
            emptySquares.insert(Square(x: index % size, y: index / size))
        }
        return emptySquares
    }
}

extension Board: CustomStringConvertible {
    var description: String {
        // Horizontal indices
        var text = "  " + self.indices.map { "\($0)" }.joined(separator: " ").bold()
        for y in self.indices {
            // Vertical index
            text += "\n\("\(y)".bold())"
            // Row values
            for x in self.indices {
                text += " \(self.board[y * self.size + x].boardCharWithColor())"
            }
        }
        return text
    }
}
